Refresh Script Canvas behavior context topic (#1182)
o3de/o3de.org: content/docs/user-guide/scripting/script-canvas/programmer-guide/behavior-context.md

+ Support for the AZ::Event and EBus event systems enable your scripts to use decoupled, event-driven programming paradigms.
+ Script Canvas can use functionality exposed through the behavior context from any Gem, enabling any Gem to enhance Script Canvas.
+ Support for Gems reflecting C++ code through the behavior context means there is no need to add Gem dependencies to Script Canvas.

## Example: Static functions

To demonstrate how C++ code can become a Script Canvas node, this example uses the behavior context to reflect a few simple, static math library functions.

We start with the static function declarations. The following functions return the sine and cosine of an angle. The angle is in radians:

```cpp
float Sin(float angle);
float Cos(float angle);
```

We also need a class that defines the namespace for these functions:

```cpp
class GlobalClass
{
public:
    AZ_TYPE_INFO(GlobalClass, "{47A07917-103F-41F5-A586-8D7C1C40A625}");
    AZ_CLASS_ALLOCATOR(GlobalClass, SystemAllocator, 0);

    GlobalClass() = default;
    ~GlobalClass() = default;
        
    static void Reflect(AZ::ReflectContext* context);
};
```

> In the class's static `Reflect` method, we use the behavior context to reflect the `GlobalClass` and bind the static `Sin` and `Cos` methods that are part of the class. In this example, the functions are configured to be part of a group called "Globals". The group is used as a subtitle on the new node and the category under which the nodes will appear in Script Canvas's Node Palette.

Author: No, you're right. I'm referring to a string in a code sample, so I should really use the code style.

In this Light component, the user can configure the light by setting parameters such as color, intensity, and radius. The Light component can also be turned on or off, and you can respond to these events when they occur. Communication with an entity's Light component is done through two EBuses: the `LightComponentRequestBus` and the `LightComponentNotificationBus`.

### Request bus

A request bus is an EBus that can send _events_. Events can be thought of as requests that are intended for a system or object to handle. Components can reflect their event methods to the behavior context to make them available to scripting environments such as Script Canvas.

Here are a few of the C++ event methods in the Light component:

```cpp
// Turn light on. Returns true if the light was successfully turned on.
bool TurnOnLight();

// Turn light off. Returns true if the light was successfully turned off.
bool TurnOffLight();

// Toggle light state.
void ToggleLight();
```

These events are part of the `LightComponentRequestBus`. Their behavior is implemented by `LightComponent`.

```cpp
bool LightComponent::TurnOnLight()
{
    bool success = m_light.TurnOn();
    if (success)
    {

>         EBUS_EVENT_ID(GetEntityId(), LightComponentNotificationBus, LightTurnedOn);

Reviewer: This style of invoking an EBUS is generally discouraged, this would change to:
`LightComponentNotificationBus::Event(GetEntityId(), &LightComponentNotifications::LightTurnedOn);`

Author: Ah, thanks! I pulled this in from legacy Light component code. Thanks for modernizing for me.

    }
    return success;
}

bool LightComponent::TurnOffLight()
{
    bool success = m_light.TurnOff();
    if (success)
    {

>         EBUS_EVENT_ID(GetEntityId(), LightComponentNotificationBus, LightTurnedOff);

Reviewer: `LightComponentNotificationBus::Event(GetEntityId(), &LightComponentNotifications::LightTurnedOff);`

Author: Fixed.

    }
    return success;
}

void LightComponent::ToggleLight()
{
    if (m_light.IsOn())
    {
        TurnOffLight();
    }
    else
    {
        TurnOnLight();
    }
}
```

To make these events accessible for scripting, their methods must be reflected to the behavior context. This is done in `LightComponent::Reflect`.

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{
    behaviorContext->EBus<LightComponentRequestBus>("Light", "LightComponentRequestBus")
        ->Attribute(AZ::Script::Attributes::Category, "Rendering")
        ->Event("TurnOn", &LightComponentRequestBus::Events::TurnOnLight)
        ->Event("TurnOff", &LightComponentRequestBus::Events::TurnOffLight)
        ->Event("Toggle", &LightComponentRequestBus::Events::ToggleLight);
}
```

When Script Canvas examines the behavior context, it finds these events and automatically generates the corresponding nodes.

![Light component nodes in Script Canvas](/images/user-guide/scripting/script-canvas/behavior-context-ebus-request-light-nodes.png)

> #### EBus events and EntityID

Author: Kind of. I'm referring to the C++ type AZ::EntityId and its appearance on Script Canvas nodes using the spelling EntityID. It's unfortunate that Script Canvas doesn't use the same case as the C++ type.

> #### EBus events and EntityID

Reviewer 2: Yeah I'd agree with keeping this as is then too.

> #### EBus events and EntityID

Author: I'll go with the actual type name spelling in the header. Good question!

> EBuses communicate with an entity's components. To do this, it needs an address. All component EBuses derive from `AZ::ComponentBus`, which is addressable by an ID of the type `AZ::EntityId`. For this reason, all nodes from a component EBus have a data pin for an `EntityID`. The presence of `Self` in the `EntityID` field refers to the `EntityID` of the entity that owns the Script Canvas graph. However, this ID can be assigned to another entity, or even changed to an invalid entity ID.

Reviewer 2: ```suggestion
EBuses communicate with an entity's components. To do this, it needs an address. All component EBuses derive from `AZ::ComponentBus`, which is addressable by an ID of the type `AZ::EntityId`. For this reason, all nodes from a component EBus have a data pin for an `EntityID`. The presence of `Self` in the `EntityID` field refers to the `EntityID` of the entity that owns the Script Canvas graph. However, this ID can be assigned to another entity, or even changed to an invalid entityId.
```
Again, not sure if a correction is appropriate here or not.

Author: Updated comment with some additional style formatting.

![Self EntityID](/images/user-guide/scripting/script-canvas/behavior-context-ebus-request-entityID.png)

#### Tooltips

Script Canvas nodes should include a helpful tooltip for every parameter. For example, the Light component might have a `SetLightState` event with a `state` parameter:

```cpp
// Set the light state to on or off.
void SetLightState(State state);
```

> You should add a tooltip in the behavior context reflection to describe the parameter. In this example, the tooltip will show "1=On, 0=Off" when a user hovers over the `State` parameter on the `SetState` node.

Reviewer 2: ```suggestion
You should add a tooltip in the behavior context reflection to describe the parameter. In this example, the tooltip will show `1=On, 0=Off` when a user hovers over the `State` parameter on the `SetState` node.
```

Author: Updated the comment to use bold for nodes and node pins and parameters. There is precedence for this in other parts of the Script Canvas section.

```cpp
    behaviorContext->EBus<LightComponentRequestBus>("Light", "LightComponentRequestBus")
        ->Event("SetState", &LightComponentRequestBus::Events::SetLightState, {{{"State", "1=On, 0=Off"}}});
```

### Notification bus

A notification bus is an EBus that enables the use of *event handlers*. Event handlers in a component can respond to the events that are sent to the component. You can reflect event handlers to the behavior context to make them available to scripting environments such as Script Canvas.

> Our Light component example handles the TurnOn, TurnOff, and Toggle events using the following C++ event handler methods on the `LightComponentNotificationBus`:

Reviewer 2: ```suggestion
Our Light component example handles the `LightTurnedOn` and `LightTurnedOff` events using the following C++ event handler methods on the `LightComponentNotificationBus`:
```

Author: I think I understand what you're getting at in this suggestion, but I'm going to leave this as is. Reason being: I wanted to refer to the original bound event, to show how TurnOn (request bus event) is handled by the following LightTurnedOn (notification bus event) event handler, for example.

Author: I'll add some clarification, though.

```cpp
if (AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
{
    ...

    behaviorContext->EBus<LightComponentNotificationBus>("LightNotification", "LightComponentNotificationBus", "Notifications for the Light components")
        ->Attribute(AZ::Script::Attributes::Category, "Rendering")
        ->Handler<BehaviorLightComponentNotificationBusHandler>();
}
```

Once compiled, these event handlers are available to Script Canvas from the `LightNotification` node.

![Light notification node](/images/user-guide/scripting/script-canvas/behavior-context-ebus-light-notification-node.png)

## Example: Data types

To make a custom data type available for use as a variable in Script Canvas, you can reflect it using the behavior context's Class builder. The new type is also available to pass as a parameter to functions and events.

This example uses the `BoxShapeConfig` class as an example. This class is defined in the file `Gems\LmbrCentral\Code\include\LmbrCentral\Shape\BoxShapeComponentBus.h` and reflected in `Gems\LmbrCentral\Code\Source\Shape\BoxShapeComponent.cpp`.

> A data type must be reflected to both the `SerializeContext` and the `BehaviorContext`. The serialization context enables the data type to be stored and read from a file, and the behavior context allows it to be bound to the scripting system.

Reviewer 2: To match the style on this page I might suggest changing this to "...reflected to both the serialization context and the behavior context.

Alternatively, this could refer to to the codeblock below and read "...reflected in both `serializeContext` and `behaviorContext`".

```cpp
void BoxShapeConfig::Reflect(AZ::ReflectContext* context)
{
    if (auto serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
    {
        serializeContext->Class<BoxShapeConfig, ShapeComponentConfig>()
            ->Version(2)
            ->Field("Dimensions", &BoxShapeConfig::m_dimensions);
     }

    if (auto behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context))
    {
         behaviorContext->Class<BoxShapeConfig>()
            ->Constructor()
            ->Constructor<AZ::Vector3&>()
            ->Property("Dimensions", BehaviorValueProperty(&BoxShapeConfig::m_dimensions));
    }
}
```

Resulting variable node:

![Get BoxShapeConfig variable node](/images/user-guide/scripting/script-canvas/behavior-context-data-types-boxshapeconfig.png)

## Best practice: Displaying EBus event parameter names in nodes

To display parameter names correctly for your EBus events, ensure that you specify custom names when you reflect your events to the behavior context.

> If you do not specify names for the parameters, they are given default display names like "1", "2", or "3", as in the following image:

Reviewer 2: ```suggestion
If you do not specify names for the parameters, they are given default display names like `1`, `2`, or `3`, as in the following image:
```
